test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -2778,77 +2781,104 @@ class test_ucp_sockaddr_protocols_err_sender
+            if (sender_idx > 0) {
+                client_ep_connect(sender_idx);
+                send_recv(sender(), sender_idx, receiver(), send_recv_type(),
+                          false, cb_type());
+            }
+
+            void *sreq = send(sender(), sender_idx, &send_buf[0], size,
+                              SEND_RECV_TAG, scomplete_cbx, NULL);
+            ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
+            ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
+
+            /* Allow receiver to get RTS notification, but do not receive message
+             * body */
+            ucp_tag_recv_info_t info;
+            ucp_tag_message_h message = message_wait(receiver(), 0, 0, &info);
+            ASSERT_NE((void*)NULL, message);
+            ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
 
-            /* Post an additional send operation for the same send buffer using
-             * another connection to test a case that the memory validation
-             * would not be completed immediately when UCP endpoint is
-             * destroyed and all UCP requests are purged */
-            second_sreq = send(second_sender(), &send_buf[0], size,
-                               SEND_RECV_TAG, scomplete_cbx, NULL);
+            send_recv_infos.emplace_back(sreq, message, nullptr, nullptr);
         }
 
         /* Ignore all errors - it is expected */
         scoped_log_handler slh(wrap_errors_logger);
-        void *sreq = send(sender(), &send_buf[0], size, SEND_RECV_TAG,
-                          scomplete_cbx, NULL);
-        ASSERT_TRUE(UCS_PTR_IS_PTR(sreq));
-        ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
-
-        /* Allow receiver to get RTS notification, but do not receive message
-         * body */
-        message = message_wait(receiver(), 0, 0, &info);
-        ASSERT_NE((void*)NULL, message);
-        ASSERT_EQ(UCS_INPROGRESS, ucp_request_check_status(sreq));
-
-        /* Close sender EP to force send operation to complete with CANCEL
-         * status */
+
+        /* Close the first sender's EP to force send operation to be completed
+         * with CANCEL status */
         entity_disconnect(sender());
 
-        if (do_extra_send) {
-            if (has_invalidate) {
+        for (auto &send_recv_info : send_recv_infos) {
+            if (has_invalidate && (num_senders > 1)) {
                 /* Check that the send request would not be completed until the
                  * memory handle for the same send buffer will not be released
                  */
+                void *sreq          = std::get<0>(send_recv_info);
                 ucs_status_t status = request_progress(sreq, { &sender() },
                                                        0.5);
                 ASSERT_EQ(UCS_INPROGRESS, status);
             }
 
-            ASSERT_TRUE(second_sreq != NULL);
+            std::get<2>(send_recv_info) = new std::string(size, 'y');
```

Comment:
test seems over-complicated now
1. why have to save the receive buffer in send_recv_info?
2. if we already had test failures, why need to change the test in order to test the fix?

Reply:
> 1. why have to save the receive buffer in send_recv_info?

to do receive into correct buffer and release it then.

> 2. if we already had test failures, why need to change the test in order to test the fix?

we didn't have test failures w/o adding the new scenario.
the test was updated to do some additional operations to not complete the UCP request for EP which is closed while some other operations are in-progress on other connections.
my first implementation of the test does `ucp_mem_map()` to prevent the UCP request from being completed.

Reply:
why not use local buffer every time we receive it?
or - pls find some other way to simplify the test

Reply:
yes, we could use local buffer for all receive operations except the first one.

Reply:
The first receive buffer is checked for invalidation error